winml/test/model/model_tests.cpp

```diff
@@ -0,0 +1,211 @@
+      WINML_EXPECT_NO_THROW(
+          actualOutput = Ort::Value::CreateTensor(
+              memoryInfo,
+              actualData,
+              actualSizeInBytes,
+              expectedShapeAndTensorType.GetShape().data(),
+              expectedShapeAndTensorType.GetShape().size(),
+              expectedShapeAndTensorType.GetElementType()));
+
+      // Use the expected and actual OrtValues to compare
+      std::pair<COMPARE_RESULT, std::string> ret = CompareOrtValue(*actualOutput, *value, m_perSampleTolerance, m_relativePerSampleTolerance, m_postProcessing);
+      WINML_EXPECT_EQUAL(COMPARE_RESULT::SUCCESS, ret.first) << ret.second;
+    }
+  }
+};
+
+TEST_P(ModelTest, Run) {
+  LearningModel model = nullptr;
+  LearningModelDevice device = nullptr;
+  LearningModelSession session = nullptr;
+  LearningModelBinding binding = nullptr;
+  WINML_EXPECT_NO_THROW(model = LearningModel::LoadFromFilePath(m_testCase->GetModelUrl()));
+  WINML_EXPECT_NO_THROW(device = LearningModelDevice(m_deviceKind));
+  WINML_EXPECT_NO_THROW(session = LearningModelSession(model, device));
+  WINML_EXPECT_NO_THROW(binding = LearningModelBinding(session));
+  for (size_t i = 0; i < m_testCase->GetDataCount(); i++) {
+    // Load and bind inputs
+    onnxruntime::test::HeapBuffer inputHolder;
+    std::unordered_map<std::string, Ort::Value> inputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, inputHolder, inputFeeds, true));
+    WINML_EXPECT_NO_THROW(BindInputsFromFeed(binding, inputFeeds));
+
+    // evaluate
+    LearningModelEvaluationResult results = nullptr;
+    WINML_EXPECT_NO_THROW(results = session.Evaluate(binding, L"Testing"));
+
+    // Load expected outputs
+    onnxruntime::test::HeapBuffer outputHolder;
+    std::unordered_map<std::string, Ort::Value> outputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, outputHolder, outputFeeds, false));
+
+    // compare results
+    CompareEvaluationResults(results, outputFeeds);
+  }
+}
+
+bool SetTestDataPath(char* testDataPath) {
+  GetEnvironmentVariableA("WINML_TEST_DATA_PATH", testDataPath, MAX_PATH);
+  if (testDataPath[0] == 0) {
+    auto hardcodedModelTestCollateralPath = FileHelpers::GetModulePath() + L"../models/";
+    printf("Need to set environment variable WINML_TEST_DATA_PATH. Trying to find hardcoded model path.. %ws", hardcodedModelTestCollateralPath.c_str());
+    if (std::filesystem::exists(hardcodedModelTestCollateralPath)) {
+      std::string hardcodedModelTestCollateralPathStr = _winml::Strings::UTF8FromUnicode(hardcodedModelTestCollateralPath.c_str(),
+                                                                                         hardcodedModelTestCollateralPath.length());
+      std::size_t length = hardcodedModelTestCollateralPathStr.copy(testDataPath, hardcodedModelTestCollateralPathStr.length());
+      testDataPath[length] = '\0';
+      return true;
+    } else {
+      return false;
+    }
+  } else {
+    return true;
+  }
```

Comment:
That's undefined behavior if `WINML_TEST_DATA_PATH` isn't set and `testDataPath` wasn't previously zero initialized, or if `WINML_TEST_DATA_PATH` has length greater than MAX_PATH (from the API doc: `If lpBuffer is not large enough to hold the data, the return value is the buffer size, in characters, required to hold the string and its terminating null character and the contents of lpBuffer are undefined`). In either of these cases `if (testDataPath[0] == 0)` is checking uninitialized memory.

There's also a buffer overflow if  `FileHelpers::GetModulePath() + L"../models/"` has length greater than MAX_PATH.